### ISSUE TITLE
Nevergrad's recommendation can give wrong result

### ISSUE THREAD

**Issue by mstimberg** [brian-team/brian2modelfitting#16]
During testing, I noticed a peculiar problem: the recommendation that we get from `Nevergrad` is not necessarily the best parameter combination that we have seen so far (at least for the differential evolution algorithm that we use by default, but I think it is more or less the same issue for the other optimizers). The reason is that `Nevergrad` always assumes that simulations are noisy, so it will not use the observed error as such. Instead, it will try to estimate lower and upper bounds on the error. This is in itself not unreasonable, even though our tests and examples all use deterministic models, so re-evaluating a parameter combination for a second time is a waste of time. But the main problem is that we can run into a numerical issue: Our errors tend to be rather small numerical values, because we use the mean squared error (and not its square root). If we are fitting currents, errors could be on the order of nA, i.e. 1e-9 and the squared error therefore on the order of 1e-18. Now, `Nevergrad` will not base its recommendation based on the observed error, but on its estimated "lower bound". This lower bound is calculated by reducing the average observed error by something like 10% of the standard error of the observed values. This should be fine if there were indeed several observed values (in a deterministic system, the standard error would be 0 of course), but when running a relatively small number of parameters in a high-dimensional space, almost all parameter combinations have only been tried once. In this case, `Nevergrad` [uses](https://github.com/facebookresearch/nevergrad/blob/1e816ead22267794344b39caf1cda75c5d6501a6/nevergrad/optimization/utils.py#L36-L37) a default variance value that is very inappropriate in our case:

```Python
        # TODO May be safer to use a default variance which depends on y for scale invariance?
        self.variance = 1.e6
```
Using this variance value that is many orders of magnitude higher than typical errors, leads to a dramatic loss in accuracy, e.g. a 10 fold difference between 1e-16 and 1e-15 will be considered the same error and which of the two gets recommended as the "best solution" is completely arbitrary. 

I could think of various ways to fix this:
* Normalize our values in some way that the absolute values are not that small (but seems to be difficult to generalize)
* for differential evolution, there's a parameter that could change the recommendation to use the "average" (i.e., measured) value instead of the calculated "optimistic" value – I did not immediately see how to set these parameters in a "clean" way, though.
* Use `float128` when calculating our errors, hoping that the type propagates to `Nevergrad`
* Do not use `Nevergrad`'s recommendation at all, but simply return the value with the lowest error (we keep track of all parameters and their errors ourselves already).

The last option is probably the easiest and least prone to break with changes in `Nevergrad`?

**Comment by romainbrette** [comment]
I think we should definitely implement the last option. But in addition, we should also add a normalization factor. The easiest I think would be to have an option in the least square metric for normalization (e.g. you would pass say 1 mV and V would divided by 1 mV). A more systematic way could be to take the average temporal standard deviation of traces, which would give an ok order of magnitude. That could be the default behavior if normalization is not explicitly provided.

**Comment by romainbrette** [comment]
I just added a normalization to MSEMetric.

**Comment by romainbrette** [comment]
Normalization does not seem make any difference in terms of error. I also noted that there is quite a bit of variability between runs, ie, they converge but to a difference value of error.

**Comment by mstimberg** [comment]
> Normalization does not seem make any difference in terms of error.

It is not directly about normalization of the error, but I just noticed that nevergrad internally uses a transformation based on the bounds we set for the individual parameters. By default, (and we don't currently allow to change it), it uses the arctan function. I think this means at least that we do not have to worry that some algorithms might not deal well with differently scaled dimensions.
